src/metabase/task/index_values.clj

```diff
@@ -0,0 +1,113 @@
+(def ^:private refreshable-states
+  "States of a model index that are refreshable."
+  #{"indexed" "initial" "error" "overflow"})
+
+(defn unindex?
+  "Whether to unindex the the model indexing job.
+  If the model does not exist, the model-index record, or if the model has reverted to a question."
+  [model model-index]
+  (or (nil? model) (nil? model-index)
+      (not (:dataset model))))
+
+(declare model-index-trigger-key)
+
+(defn- refresh-index!
+  "Function "
+  [model-index-id]
+  (let [model-index              (t2/select-one ModelIndex :id model-index-id)
+        model                    (when model-index
+                                   (t2/select-one Card :id (:model_id model-index)))]
+    (if (unindex? model model-index)
+      (u/ignore-exceptions
+       (let [trigger-key (model-index-trigger-key model-index-id)]
+         (task/delete-trigger! trigger-key)))
+      (model-index/add-values! model-index))))
+
+(jobs/defjob ^{org.quartz.DisallowConcurrentExecution true
+               :doc "Refresh model indexed columns"}
+  ModelIndexRefresh [job-context]
+  (let [{:strs [model-index-id]} (qc/from-job-data job-context)]
+    (refresh-index! model-index-id)))
+
+(def ^:private refresh-model-index-key
+  "Job key string for refresh job. Call `(jobs/key refresh-model-index-key)` if you need the org.quartz.JobKey
+  instance."
+  "metabase.task.IndexValues.job")
+
+(def ^:private refresh-job
+  (jobs/build
+   (jobs/with-description "Indexed Value Refresh task")
+   (jobs/of-type ModelIndexRefresh)
+   (jobs/with-identity (jobs/key refresh-model-index-key))
+   (jobs/store-durably)))
+
+(defn- model-index-trigger-key
+  [model-index-id]
+  (triggers/key
+   (format "metabase.task.IndexValues.trigger.%d" model-index-id)))
+
+(defn- refresh-trigger ^org.quartz.CronTrigger [model-index]
+  (triggers/build
+   (triggers/with-description (format "Refresh index on model %d" (:model_id model-index)))
+   (triggers/with-identity (model-index-trigger-key (:id model-index)))
+   (triggers/using-job-data {"model-index-id" (u/the-id model-index)})
+   (triggers/for-job (jobs/key refresh-model-index-key))
+   (triggers/start-now)
+   (triggers/with-schedule
+     (cron/schedule
+      (cron/cron-schedule (:schedule model-index))
+      (cron/in-time-zone (TimeZone/getTimeZone (or (driver/report-timezone)
+                                                   (qp.timezone/system-timezone-id)
+                                                   "UTC")))
+      (cron/with-misfire-handling-instruction-do-nothing)))))
+
+(defn add-indexing-job
+  "Public API to start indexing a model."
+  [model-index]
+  (let [trigger (refresh-trigger model-index)]
+    (log/info
+     (u/format-color :green (trs "Scheduling indexing for model: {0}" (:model_id model-index))))
+    (try (task/add-trigger! trigger)
+         (catch ObjectAlreadyExistsException _e
+           (log/info (u/format-color :green (trs "Persistence already present for model: {0}"
```

Comment:
```suggestion
           (log/info (u/format-color :green (trs "Persistence already present for model: {0}"
```
this error message seems off? did you mean the model index already present?

Reply:
copy/paste error. thanks